drivers/timer/stm32_lptim_timer.c

```diff
@@ -0,0 +1,270 @@
+#define LPTIM_TIMEBASE 0xF9FF
+#elif (LPTIM_CLOCK == LSE_VALUE)
+#define LPTIM_TIMEBASE 0xFFFF
+#else
+#error "LPTIM_CLOCK not valid"
+#endif
+#define LPTIM_MAX_COUNT 0xFFFF
+#define LPTIM_PRESCALER 2
+
+/* time in ms of one LPTIM counter unit */
+#define LPTIM_COUNT_UNIT ((LPTIM_PRESCALER * 1000) / LPTIM_CLOCK)
+
+/* nb of LPTIM counter unit per ms */
+#define COUNT_PER_MS (LPTIM_CLOCK / (LPTIM_PRESCALER * 1000))
+
+/* assuming one TICKS is 10 ms long */
+#define MS_PER_TICK (1000 / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+/* nb of LPTIM counter unit per kernel tick in ms (MS_PER_TICK) */
+#define COUNT_PER_TICK (MS_PER_TICK * LPTIM_CLOCK \
+		      / (LPTIM_PRESCALER * 1000))
+
+struct device;	       /* forward declaration; type is not used. */
+
+static volatile u32_t last_cnt;
+
+static struct k_spinlock lock;
+
+
+static void lptim_irq_handler(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	if ((LL_LPTIM_IsActiveFlag_ARRM(LPTIM1) != 0)
+		&& (LL_LPTIM_IsEnabledIT_ARRM)(LPTIM1) != 0) {
+
+		k_spinlock_key_t key = k_spin_lock(&lock);
+
+		/* LPTIM1 CNT could have already been reset */
+		u32_t cnt_value = LL_LPTIM_GetAutoReload(LPTIM1) + 1;
+
+		/* next counter reload match Interrupt */
+		LL_LPTIM_SetAutoReload(LPTIM1, LPTIM_TIMEBASE);
+		/* if needed, a next timeout will be raised */
+		LL_LPTIM_ClearFLAG_ARRM(LPTIM1);
+		/* reset last_cnt for a next call to 'elapsed' */
+		last_cnt = 0;
+
+		k_spin_unlock(&lock, key);
+
+		/* announce the elapsed time in ms. */
+		u32_t dticks = z_ms_to_ticks(((cnt_value *
+			LPTIM_PRESCALER * 1000) / LPTIM_CLOCK)) + 1;
+		z_clock_announce(IS_ENABLED(CONFIG_TICKLESS_KERNEL)
+			? dticks : 1);
+	}
+}
+
+int z_clock_driver_init(struct device *device)
+{
+	ARG_UNUSED(device);
+
+	/* enable LPTIM clock source */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
+	LL_APB1_GRP1_ReleaseReset(LL_APB1_GRP1_PERIPH_LPTIM1);
+	/* select LSI as LPTIM1 clock source */
+#if (LPTIM_CLOCK == LSI_VALUE)
+	/* enable LSI clock */
+	LL_RCC_LSI_Enable();
+	while (!LL_RCC_LSI_IsReady()) {
+		/* Wait for LSI ready */
+	}
+	LL_RCC_SetLPTIMClockSource(LL_RCC_LPTIM1_CLKSOURCE_LSI);
+#elif (LPTIM_CLOCK == LSE_VALUE)
+
+	/* Enable the power interface clock */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+
+	/* enable backup domain */
+	LL_PWR_EnableBkUpAccess();
+	LL_RCC_ForceBackupDomainReset();
+	LL_RCC_ReleaseBackupDomainReset();
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) &&	\
+	!defined(CONFIG_SOC_SERIES_STM32F2X)
+
+	LL_RCC_LSE_SetDriveCapability(LL_RCC_LSEDRIVE_MEDIUMHIGH);
+
+#endif /* !CONFIG_SOC_SERIES_STM32F4X && !CONFIG_SOC_SERIES_STM32F2X */
+	/* enable LSE clock */
+	LL_RCC_LSE_DisableBypass();
+	LL_RCC_LSE_Enable();
+	while (!LL_RCC_LSE_IsReady()) {
+		/* Wait for LSE ready */
+	}
+	LL_RCC_SetLPTIMClockSource(LL_RCC_LPTIM1_CLKSOURCE_LSE);
+#else
+#error "LPTIM_CLOCK not valid"
+#endif
+
+	/* Clear the event flag and possible pending interrupt */
+	IRQ_CONNECT(LPTIM1_IRQn, 1, lptim_irq_handler, 0, 0);
```

Comment:
Please use data from device tree.

Reply:
LPTIM1_IRQn is DT_LPTIM_1_IRQ defined by the DT_ST_STM32_TIMERS_40007C00_IRQ_0 in the dts